utils/python/CIME/macros.py

```diff
@@ -0,0 +1,1007 @@
+"""
+Classes used to build the CIME Macros file.
+
+The main "public" class here is MacroMaker. Other classes are mainly for
+internal use, though the various "MacroWriter" classes might be useful
+elsewhere one day.
```

Comment:
This is a lot of new code and appears to be a fairly major change. Is there any way we could get a little more documentation here? What is the design? Why is it better than before?